libtransact/src/contract/handler.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright 2019 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::hash::Hash;
+
+use crate::contract::address::Addresser;
+use crate::handler::{ApplyError, TransactionContext, TransactionHandler};
+use crate::protocol::transaction::TransactionPair;
+
+pub trait SmartContract: Send {
+    type Key: Eq + Hash;
+    type Addr: Addresser<Self::Key>;
+    type Context;
+
+    fn get_family_name(&self) -> &str;
+
+    fn get_family_versions(&self) -> &[String];
+
+    fn get_addresser(&self) -> Self::Addr;
+
+    fn make_context<'a>(
+        &self,
+        addresser: Self::Addr,
+        context: &'a mut dyn TransactionContext,
+    ) -> Self::Context;
+
+    fn apply(
+        &self,
+        transaction: &TransactionPair,
+        context: Self::Context,
+    ) -> Result<(), ApplyError>;
+}
+
+impl<T> TransactionHandler for T
```

Comment:
This should be removed from here (this code shouldn't know about TransactionHandler concepts). To mesh these together without changing the TransactionHandler API (yet), we could create a SmartContractEngine trait, impl SmartContract on that trait, and then create a SmartContractEngineTransactionHandler implementation which pulls it together. I think longer-term, we get rid of TransactionHandler in favor of SmartContractEngine, but this approach would buy us some space to figure out how SmartContractEngine should be defined.